[OMPIRBuilder] always leave PARALLEL via the same barrier

tblah opened the pull request:
A barrier will pause execution until all threads reach it. If some go to a different barrier then we deadlock. This manifests in that the finalization callback must only be run once. Fix by ensuring we always go through the same finalization block whether the thread in cancelled or not and no matter which cancellation point causes the cancellation.

The old callback only affected PARALLEL, so it has been moved into the code generating PARALLEL. For this reason, we don't need similar changes for other cancellable constructs. We need to create the barrier on the shared exit from the outlined function instead of only on the cancelled branch to make sure that threads exiting normally (without cancellation) meet the same barriers as those which were cancelled. For example, previously we might have generated code like

```
...
  %ret = call i32 @__kmpc_cancel(...)
  %cond = icmp eq i32 %ret, 0
  br i1 %cond, label %continue, label %cancel

continue:
  // do the rest of the callback, eventually branching to %fini
  br label %fini

cancel:
  // Populated by the callback:
  // unsafe: if any thread makes it to the end without being cancelled
  // it won't reach this barrier and then the program will deadlock
  %unused = call i32 @__kmpc_cancel_barrier(...)
  br label %fini

fini:
  // run destructors etc
  ret
```

In the new version the barrier is moved into fini. I generate it *after* the destructors because the standard describes the barrier as occurring after the end of the parallel region.

```
...
  %ret = call i32 @__kmpc_cancel(...)
  %cond = icmp eq i32 %ret, 0
  br i1 %cond, label %continue, label %cancel

continue:
  // do the rest of the callback, eventually branching to %fini
  br label %fini

cancel:
  br label %fini

fini:
  // run destructors etc
  // safe so long as every exit from the function happens via this block:
  %unused = call i32 @__kmpc_cancel_barrier(...)
  ret
```

To achieve this, the barrier is now generated alongside the finalization code instead of in the callback. This is the reason for the changes to the unit test.

I'm unsure if I should keep the incorrect barrier generation callback only on the cancellation branch in clang with the OMPIRBuilder backend because that would match clang's ordinary codegen. Right now I have opted to remove it entirely because it is a deadlock waiting to happen.

tblah committed:
[OMPIRBuilder] always leave PARALLEL via the same barrier

A barrier will pause execution until all threads reach it. If some go to
a different barrier then we deadlock. This manifests in that the
finalization callback must only be run once. Fix by ensuring we always
go through the same finalization block whether the thread in cancelled
or not and no matter which cancellation point causes the cancellation.

The old callback only affected PARALLEL, so it has been moved into the
code generating PARALLEL. For this reason, we don't need similar changes
for other cancellable constructs. We need to create the barrier on the
shared exit from the outlined function instead of only on the cancelled
branch to make sure that threads exiting normally (without cancellation)
meet the same barriers as those which were cancelled. For example,
previously we might have generated code like

```
...
  %ret = call i32 @__kmpc_cancel(...)
  %cond = icmp eq i32 %ret, 0
  br i1 %cond, label %continue, label %cancel

continue:
  // do the rest of the callback, eventually branching to %fini
  br label %fini

cancel:
  // Populated by the callback:
  // unsafe: if any thread makes it to the end without being cancelled
  // it won't reach this barrier and then the program will deadlock
  %unused = call i32 @__kmpc_cancel_barrier(...)
  br label %fini

fini:
  // run destructors etc
  ret
```

In the new version the barrier is moved into fini. I generate it *after*
the destructors because the standard describes the barrier as occurring
after the end of the parallel region.

```
...
  %ret = call i32 @__kmpc_cancel(...)
  %cond = icmp eq i32 %ret, 0
  br i1 %cond, label %continue, label %cancel

continue:
  // do the rest of the callback, eventually branching to %fini
  br label %fini

cancel:
  br label %fini

fini:
  // run destructors etc
  // safe so long as every exit from the function happens via this block:
  %unused = call i32 @__kmpc_cancel_barrier(...)
  ret
```

To achieve this, the barrier is now generated alongside the finalization
code instead of in the callback. This is the reason for the changes to
the unit test.

I'm unsure if I should keep the incorrect barrier generation callback only
on the cancellation branch in clang with the OMPIRBuilder backend
because that would match clang's ordinary codegen. Right now I have
opted to remove it entirely because it is a deadlock waiting to happen.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1800,8 +1779,18 @@ OpenMPIRBuilder::InsertPointOrErrorTy OpenMPIRBuilder::createParallel(
-  if (Error Err = FiniCB(PreFiniIP))
-    return Err;
+  if (!FiniInfo.FiniBB) {
+    if (Error Err = FiniCB(PreFiniIP))
```
What would set `FiniInfo.FiniBB` after the callback?

If it is a call to `createBarrier` that the `FiniCB`, that's the kind of devil's contract that make the callback-driven design so bad.

tblah replied:
The finalisation basic block could have already been created if the body of the parallel operation contained a cancellation point or cancel. In that case we should just branch straight to the block created previously. I agree the control flow with all of the callbacks and the cancellation stack are a bit hard to follow. This is not new with this patch.

In most cases, no cancellation will have already created a finalisation block so finalisation should be generated right here as was done before this patch.

The intention here is to only run the finalisation callback once and have all exists branch to that one instance (and also to include the barrier in that unique exit so that all threads block on the same barrier).

It is not expected that the FiniCB creates FiniBB.

Meinersbur replied:
Can we improve the situation a bit to make it more predictable? There seem to be two code locations where FiniCB is called and FiniBB is assigned, here or in `commonDirectiveExit`. I propose refactoring this out into a different function whose task it is to get the FiniBB: If it was already created, return it; otherwise emit one. Such as

```cpp
struct FinalizationInfo {
  ...
  BasicBlock *getOrCreateFinializationBranchTarget() {
    if (!FiniBB) {
     FiniCB(FiniIP);
     FiniBB = FinIP.getBlock();
    }
    return FiniBB;
  }
};
```

tblah committed:
Remove redundant use of llvm namespace

bhandarkar-pranav reviewed (approved):
LGTM, thank you for the PR, @tblah

ergawy commented:
I have a newbie (not familiar with the semantics of the `cancel` construct) question specially given the changes in this PR, what is the point of emitting the `__kmpc_cancel_barrier` specially that we now emit it at the very end of the parallel region?

tblah commented:
@Meinersbur, @ergawy are you happy with this patch?

Meinersbur reviewed (changes requested):
I think the possible reuse of finialization-unrelated code in FinIP.getBlock() should be resolved before landing this PR

tblah committed:
Move finalization block creation to helpers

Justifications for where it is safe to re-use an existing block for
finalization:
  - omp sections: the finalization block here is the one already
    containing the loop finalization code. The structure of wsloop
    codegen ensures this only ever contains loop finalization code, and
    the existing cancellation implementation already relied upon this.
  - emitCommonDirectiveExit: the block is created afresh especially for
    this call. I got rid of the merge with the predecessor block. This
    leads to most of the test updates.

This simplifies omp sections code slightly because we handle branches
better than before and don't need to create a dummy that is then touched
up.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -675,6 +675,60 @@ OpenMPIRBuilder::getOrCreateRuntimeFunction(Module &M, RuntimeFunction FnID) {
+    }
+    I.insertAfter(InsertAfter);
+    InsertAfter = BasicBlock::iterator(&I);
+  }
```
Use `BasicBlock::splice`?

Meinersbur reviewed (commented):
For avoiding emitting finalization code multiple times, Clang generates code like this:
```
ExitingBB1:
  br label %FiniBB

Exiting BB2:
  br label %FiniBB

FiniBB:
  %origin = phi i32 [%ExitingBB1, 0], [%ExitingBB2, 1]
  ... finalization code (calling dtors etc.) ...
  switch i32 %origin
     [0, label %ExitBB1],
     [1, label %ExitBB2]

ExitBB1:
  ...

ExitBB2:
  ...
```

If I read the code correctly, `mergeFiniBB` merges all previous FiniBBs into the new OtherFiniBB (why not make the new code use the existing pre-FiniBB by emitting a branch to it?), and therefore effectively requires that all FiniBBs branch to the the same exit. If so, could you add a comment for that (in addition to "must not contain any non-finalization code")?

tblah committed:
Use BasicBlock::splice

tblah committed:
Use early exits

tblah committed:
Use Value::replaceAllUsesWith

Meinersbur reviewed (approved):
LGTM, thank you

tblah commented:
I think this caused a buildbot failure: https://lab.llvm.org/buildbot/#/builders/169/builds/17519/steps/11/logs/stdio
```
    #12 0x5e004e0e3d51 in llvm::OpenMPIRBuilder::FinalizationInfo::mergeFiniBB(llvm::IRBuilderBase&, llvm::BasicBlock*) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:706:21
```
